asv_benchmarks/benchmarks/common.py

```diff
@@ -0,0 +1,223 @@
+        data_size = 'large'
+
+    @property
+    @classmethod
+    @abstractmethod
+    def params(self):
+        pass
+
+
+class Estimator(ABC):
+    @abstractmethod
+    def setup_cache_(self, params):
+        pass
+
+    def setup_cache(self):
+        clear_tmp()
+
+        param_grid = list(itertools.product(*self.params))
+
+        for params in param_grid:
+            data, estimator = self.setup_cache_(params) or (None, None)
+            if data is None:
+                continue
+
+            data_path = get_data_path(self, params)
+            with open(data_path, 'wb') as f:
+                pickle.dump(data, f)
+
+            X, _, y, _ = data
+            estimator.fit(X, y)
+
+            est_path = get_estimator_path(self, Benchmark.save_folder,
+                                          params, Benchmark.save_estimators)
+            with open(est_path, 'wb') as f:
+                pickle.dump(estimator, f)
+
+    def setup(self, *params):
+        if hasattr(self, 'setup_'):
+            self.setup_(params)
+
+        data_path = get_data_path(self, params)
+        with open(data_path, 'rb') as f:
+            self.X, self.X_val, self.y, self.y_val = pickle.load(f)
+
+        est_path = get_estimator_path(self, Benchmark.save_folder,
+                                      params, Benchmark.save_estimators)
+        with open(est_path, 'rb') as f:
+            self.estimator = pickle.load(f)
+
+        self.make_scorers()
+
+    def time_fit(self, *args):
+        self.estimator.fit(self.X, self.y)
+
+    def peakmem_fit(self, *args):
+        self.estimator.fit(self.X, self.y)
+
+    def track_train_score(self, *args):
+        if hasattr(self.estimator, 'predict'):
+            y_pred = self.estimator.predict(self.X)
+        else:
+            y_pred = None
+        return float(self.train_scorer(self.y, y_pred))
+
+    def track_test_score(self, *args):
+        if hasattr(self.estimator, 'predict'):
+            y_val_pred = self.estimator.predict(self.X_val)
+        else:
+            y_val_pred = None
+        return float(self.test_scorer(self.y_val, y_val_pred))
+
+
+class Predictor(ABC):
+    if Benchmark.bench_predict:
+        def time_predict(self, *args):
+            self.estimator.predict(self.X)
+
+        def peakmem_predict(self, *args):
+            self.estimator.predict(self.X)
+
+        if Benchmark.base_folder is not None:
+            def track_same_prediction(self, *args):
+                file_path = get_estimator_path(self, Benchmark.base_folder,
+                                               args, True)
+                with open(file_path, 'rb') as f:
+                    estimator_base = pickle.load(f)
+
+                y_val_pred_base = estimator_base.predict(self.X_val)
+                y_val_pred = self.estimator.predict(self.X_val)
+
+                return np.allclose(y_val_pred_base, y_val_pred)
+
+    @property
+    @classmethod
+    @abstractmethod
+    def params(self):
```

Comment:
Are all 3 necessary? How can it even work: I would have imagined property and classmethod have conflicting signatures?

Reply:
apparently it works :) but you're right only 2 are necessary. I changed that.